pallets/eth-bridge/src/tests/tests.rs

```diff
@@ -90,3 +94,164 @@ fn run_checks(
+    use frame_system::RawOrigin;
+
+    fn run_eth_tx_hash_test(setup_fn: Option<fn(&mut ActiveTransactionData<TestRuntime>)>, expected: DispatchResultWithPostInfo) {
+        let mut ext = ExtBuilder::build_default().with_validators().as_externality();
+        ext.execute_with(|| {
+            let context = setup_context();
+
+            let tx_id = add_new_request::<TestRuntime>(
+                &context.request_function_name,
+                &context.request_params,
+            )
+            .unwrap();
+
+            if let Some(setup_fn) = setup_fn {
+                let mut active_tx = ActiveTransaction::<TestRuntime>::get().expect("is active");
+                setup_fn(&mut active_tx);
+                ActiveTransaction::<TestRuntime>::put(active_tx);
+            }
+
+            let result = EthBridge::add_eth_tx_hash(
+                RawOrigin::None.into(),
+                tx_id,
+                context.eth_tx_hash,
+                context.author,
+                context.test_signature.clone()
+            );
+
+            assert_eq!(result, expected);
+        });
+    }
+
+    #[test]
+    fn it_returns_error_when_eth_tx_hash_already_set() {
+        run_eth_tx_hash_test(Some(|active_tx| {
+            active_tx.data.eth_tx_hash = H256::repeat_byte(1);
+        }), Err(Error::<TestRuntime>::EthTxHashAlreadySet.into()));
+    }
+
+    #[test]
+    fn it_returns_error_when_eth_tx_hash_must_be_set_by_sender() {
+        run_eth_tx_hash_test(Some(|active_tx| {
+            active_tx.data.sender = Default::default();
+        }), Err(Error::<TestRuntime>::EthTxHashMustBeSetBySender.into()));
+    }
+
+    #[test]
+    fn it_sets_eth_tx_hash_correctly() {
+        run_eth_tx_hash_test(None, Ok(().into()));
+    }
+}
+
+#[cfg(test)]
+mod add_corroboration {
+    use super::*;
+    use frame_support::assert_ok;
+    use frame_system::RawOrigin;
+    fn run_corroboration_test(
+        is_tx_successful: bool,
+        is_hash_valid: bool,
+        assertion: fn(ActiveTransactionData<TestRuntime>),
+    ) {
+        let mut ext = ExtBuilder::build_default().with_validators().as_externality();
+        ext.execute_with(|| {
+            let context = setup_context();
+
+            let tx_id = add_new_request::<TestRuntime>(
+                &context.request_function_name,
+                &context.request_params,
+            )
+            .unwrap();
+
+            assert_ok!(EthBridge::add_corroboration(
+                RawOrigin::None.into(),
+                tx_id,
+                is_tx_successful,
+                is_hash_valid,
+                context.confirming_author.clone(),
+                context.test_signature
+            ));
+
+            let active_tx = ActiveTransaction::<TestRuntime>::get()
+                .expect("Active transaction should be present");
+
+            assertion(active_tx);
+        });
+    }
+
+
+
+    #[test]
+    fn it_adds_invalid_hash_and_successful_corroboration_correctly() {
```

Comment:
I find these tests difficult to see what they are doing. IMO there is too much abstraction here.

Reply:
I agree, in the sense that the whole test is encapsulated in a single function and the different types of tests are specified in the arguments.

That is nifty in terms of coding but overall reduces the value of the test because it makes them obfuscated. In tests I prefer dumb code, even to the point of repeating a few actions instead of looping. 

But I also like to capture elementary steps inside well-named functions that describe what they do. This is abstraction as well, but of a different kind: it focuses the test to stay as much as possible _at a single level of abstraction_.